src/test/java/org/mifos/integrationtest/cucumber/stepdef/BillPayStepDef.java

```diff
@@ -309,4 +323,195 @@ public void iCanExtractTheCallbackBodyAndAssertTheRtpStatus() {
+                        responseDescription = rootNode.get("responseDescription").asText();
+                    }
+                    assertThat(responseDescription).isNotEmpty();
+                    assertThat(responseDescription).contains("Unindentified Biller");
+                    String responseCode = null;
+                    if (rootNode.has("responseCode")) {
+                        responseCode = rootNode.get("responseCode").asText();
+                    }
+                    assertThat(responseCode).isNotEmpty();
+                }
+            }
+
+        }
+        assertThat(flag).isTrue();
+    }
+
+    @Then("I should be able to extract response body from callback for bill invalid")
+    public void iShouldBeAbleToExtractResponseBodyFromCallbackForBillInvalid() {
+        boolean flag = false;
+        List<ServeEvent> allServeEvents = getAllServeEvents();
+        for (int i = allServeEvents.size() - 1; i >= 0; i--) {
+            ServeEvent request = allServeEvents.get(i);
+            if (!(request.getRequest().getBodyAsString()).isEmpty()) {
+                JsonNode rootNode = null;
+                flag = true;
+                try {
+                    rootNode = objectMapper.readTree(request.getRequest().getBody());
+                    logger.info("Rootnode value:" + rootNode);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+                if (rootNode != null && rootNode.has("code") && rootNode.get("code").asText().equals("01")) {
+                    String reason = null;
+                    if (rootNode.has("reason")) {
+                        reason = rootNode.get("reason").asText();
+                    }
+                    assertThat(reason).isNotEmpty();
+                    assertThat(reason).contains("Invalid Bill ID");
+                    String code = null;
+                    if (rootNode.has("code")) {
+                        code = rootNode.get("code").asText();
+                    }
+                    assertThat(code).isNotEmpty();
+                }
+            }
+
+        }
+        assertThat(flag).isTrue();
+    }
+
+    @And("I should get Payer FSP not found in response")
+    public void iShouldGetDataInResponse() throws JSONException {
+        JSONObject jsonObject = new JSONObject(scenarioScopeState.response);
+        scenarioScopeState.transactionId = jsonObject.getString("transactionId");
+        assertThat(scenarioScopeState.transactionId.equals("Participant Not Onboarded")).isTrue();
+    }
+
+    @Then("I should be able to extract response body from callback for empty bill id")
+    public void iShouldBeAbleToExtractResponseBodyFromCallbackForEmptyBillId() {
+        boolean flag = false;
+        List<ServeEvent> allServeEvents = getAllServeEvents();
+        for (int i = allServeEvents.size() - 1; i >= 0; i--) {
+            ServeEvent request = allServeEvents.get(i);
```

Comment:
Looks like this code is repeated all over, can we use a single method for finding in events and reuse them?